File: spring-boot-s3-java-maven-aws-java-sdk/src/main/java/io/pivotal/tsalm/s3/S3Resource.java
package io.pivotal.tsalm.s3;


import io.pivotal.tsalm.s3.infrastructure.S3KeyDoesNotExistException;
import io.pivotal.tsalm.s3.infrastructure.S3Service;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.multipart.MultipartHttpServletRequest;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Optional;

@RestController
@RequestMapping(S3Resource.BASE_URI)
public class S3Resource {

    static final String BASE_URI = "/api/v1/buckets";

    private S3Service service;

    @Autowired
    public S3Resource(S3Service service) {
        this.service = service;
    }

    @GetMapping("/{bucketName}/objects/{objectKey}")
    public ResponseEntity<Resource> downloadS3Object(@PathVariable String bucketName, @PathVariable String objectKey) {
        try {
            final Resource s3Object = service.getObject(bucketName, objectKey);
            return ResponseEntity.ok(s3Object);
        } catch (S3KeyDoesNotExistException exception) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping(path = "/{bucketName}/objects", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<Void> uploadS3Object(@PathVariable String bucketName, MultipartHttpServletRequest request)
            throws IOException, URISyntaxException {
        final Optional<MultipartFile> multipartFile = request.getMultiFileMap().toSingleValueMap().values().stream().findFirst();
        if (multipartFile.isEmpty()) {
            return ResponseEntity.badRequest().build();
        }

        final String objectKey = service.putObject(bucketName, multipartFile.get().getResource());
        final URI location = new URI(String.format("%s/%s/objects/%s", S3Resource.BASE_URI, bucketName, objectKey));
        return ResponseEntity.created(location).build();
    }

    @DeleteMapping("/{bucketName}/objects/{objectKey}")
    public ResponseEntity<Void> removeObject(@PathVariable String bucketName, @PathVariable String objectKey) {
        try {
            service.removeObject(bucketName, objectKey);
            return ResponseEntity.noContent().build();
        } catch (S3KeyDoesNotExistException exception) {
            return ResponseEntity.notFound().build();
        }
    }
}

File: spring-boot-s3-kotlin-gradle-aws-java-sdk/src/main/kotlin/io/pivotal/tsalm/s3/S3Resource.kt
package io.pivotal.tsalm.s3


import io.pivotal.tsalm.s3.infrastructure.S3KeyDoesNotExistException
import io.pivotal.tsalm.s3.infrastructure.S3Service
import org.springframework.core.io.Resource
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartHttpServletRequest
import java.net.URI


@RestController
@RequestMapping(S3Resource.BASE_URI)
class S3Resource(private val service: S3Service) {

    companion object {
        const val BASE_URI = "/api/v1/buckets"
    }

    @GetMapping("/{bucketName}/objects/{objectKey}")
    fun downloadS3Object(@PathVariable bucketName: String, @PathVariable objectKey: String): ResponseEntity<Resource> {
        return try {
            val s3Object = service.getObject(bucketName, objectKey)
            ResponseEntity.ok(s3Object)
        } catch (exception: S3KeyDoesNotExistException) {
            ResponseEntity.notFound().build()
        }
    }

    @PostMapping("/{bucketName}/objects", consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    fun uploadS3Object(@PathVariable bucketName: String, request: MultipartHttpServletRequest): ResponseEntity<Void> {
        val multipartFile = request.multiFileMap.toSingleValueMap().values.first()
        val objectKey = service.putObject(bucketName, multipartFile.resource)
        return ResponseEntity.created(URI("$BASE_URI/$bucketName/objects/$objectKey")).build()
    }

    @DeleteMapping("/{bucketName}/objects/{objectKey}")
    fun removeObject(@PathVariable bucketName: String, @PathVariable objectKey: String): ResponseEntity<Void> {
        return try {
            service.removeObject(bucketName, objectKey)
            return ResponseEntity.noContent().build()
        } catch (exception: S3KeyDoesNotExistException) {
            ResponseEntity.notFound().build()
        }
    }
}